test-tests/shared/src/test/scala/zio/test/environment/RandomSpec.scala

```diff
@@ -0,0 +1,168 @@
+    val seed    = SRandom.nextLong()
+    val sRandom = new SRandom(seed)
+    for {
+      testRandom <- TestRandom.makeTest(DefaultData)
+      _          <- testRandom.setSeed(seed)
+      actual     <- UIO.foreach(List.fill(100)(()))(_ => f(testRandom))
+      expected   = List.fill(100)(g(sRandom))
+    } yield assert(actual, equalTo(expected))
+  }
+
+  def forAllEqualBytes: ZIO[Any, Nothing, TestResult] = {
+    val seed    = SRandom.nextLong()
+    val sRandom = new SRandom(seed)
+    for {
+      testRandom <- TestRandom.makeTest(DefaultData)
+      _          <- testRandom.setSeed(seed)
+      actual     <- UIO.foreach(List.range(0, 100))(testRandom.nextBytes(_))
+      expected = List.range(0, 100).map(new Array[Byte](_)).map { arr =>
+        sRandom.nextBytes(arr)
+        Chunk.fromArray(arr)
+      }
+    } yield assert(actual, equalTo(expected))
+  }
+
+  def forAllEqualGaussian: ZIO[Any, Nothing, TestResult] = {
+    val seed    = SRandom.nextLong()
+    val sRandom = new SRandom(seed)
+    for {
+      testRandom <- TestRandom.makeTest(DefaultData)
+      _          <- testRandom.setSeed(seed)
+      actual     <- UIO.foreach(List.fill(100)(()))(_ => testRandom.nextGaussian)
+      expected   = List.fill(100)(sRandom.nextGaussian)
+    } yield assert(actual.zip(expected).forall { case (x, y) => math.abs(x - y) < 0.01 }, equalTo(true))
+  }
+
+  def forAllEqualN[A](f: (Test, Int) => UIO[A])(g: (SRandom, Int) => A): ZIO[Any, Nothing, TestResult] = {
+    val seed    = SRandom.nextLong()
+    val sRandom = new SRandom(seed)
+    for {
+      testRandom <- TestRandom.makeTest(DefaultData)
+      _          <- testRandom.setSeed(seed)
+      actual     <- UIO.foreach(1 to 100)(f(testRandom, _))
+      expected   = (1 to 100).map(g(sRandom, _))
+    } yield assert(actual, equalTo(expected))
+  }
+
+  def forAllEqualShuffle(
+    f: (Test, List[Int]) => UIO[List[Int]]
+  )(g: (SRandom, List[Int]) => List[Int]): ZIO[Any, Nothing, TestResult] = {
+    val seed    = SRandom.nextLong()
+    val sRandom = new SRandom(seed)
+    for {
+      testRandom <- TestRandom.makeTest(DefaultData)
+      _          <- testRandom.setSeed(seed)
+      actual     <- UIO.foreach(List.range(0, 100).map(List.range(0, _)))(f(testRandom, _))
+      expected   = List.range(0, 100).map(List.range(0, _)).map(g(sRandom, _))
+    } yield assert(actual, equalTo(expected))
+  }
+
+  def forAllBounded[A: Numeric](bound: SRandom => A)(f: (Test, A) => UIO[A]): ZIO[Any, Nothing, TestResult] = {
+    val num = implicitly[Numeric[A]]
+    import num._
+    val seed    = SRandom.nextLong()
+    val sRandom = new SRandom(seed)
+    for {
+      testRandom <- TestRandom.makeTest(DefaultData)
+      _          <- testRandom.setSeed(seed)
+      bounds     = List.fill(100)(num.abs(bound(sRandom)) max one)
+      actual     <- UIO.foreach(bounds)(f(testRandom, _))
+    } yield assert(actual.zip(bounds).forall { case (a, n) => zero <= a && a < n }, equalTo(true))
```

Comment:
If you want to have a more specific assertion here, you are on the right track with `assert(actual.zip(bounds), forall(...)`. The challenge is that `within` is not going to work as is because `isWithin` takes two parameters for the bounds and is expecting an `A`, but you need an `Assertion[(A, B)]`. You need a combinator that, given an `Assertion[A]` and a way to convert an `(A, B)` to an `A` and an `Assertion[A]`, gives you an `Assertion[(A, B)]`. The closest thing we have to this is `AssertionRec` but it is not quite powerful enough because it does not let you generate the assertion dynamically.

I will think about this for the future but I think this is fine for now.